op-batcher/batcher/espresso.go

```diff
@@ -0,0 +1,163 @@
+package batcher
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	espressoVerification "github.com/EspressoSystems/espresso-sequencer-go/verification"
+
+	espressoCommon "github.com/EspressoSystems/espresso-sequencer-go/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// TODO: Pull out to be re-used in op-node for derivation from Espresso
+type Transaction struct {
+	// Namespace of transaction to be published
+	Namespace uint64
+	// TEE attestation to be verified by op-node
+	TeeAttn []byte
+	// Frames serialized as they would be for posting to L1 as calldata
+	CallData []byte
+}
+
+func (t Transaction) toEspresso() espressoCommon.Transaction {
+	payload := append(t.TeeAttn, t.CallData...)
+	return espressoCommon.Transaction{
+		Namespace: t.Namespace,
+		Payload:   payload,
+	}
+}
+
+func (l *BatchSubmitter) waitForFinality(height uint64, rawHeader json.RawMessage, header *espressoCommon.HeaderImpl) error {
+	timer := time.NewTimer(2 * time.Minute)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	var snapshot espressoCommon.BlockMerkleSnapshot
+
+Loop:
+	for {
+		select {
+		case <-ticker.C:
+			res, err := l.EspressoLightClient.FetchMerkleRoot(height, nil)
+			if err == nil {
+				snapshot = res
+				break Loop
+			}
+		case <-timer.C:
+			return fmt.Errorf("failed to fetch merkle root")
+		}
+	}
+
+	if snapshot.Height <= height {
+		return fmt.Errorf("snapshot height is less than or equal to the requested height")
```

Comment:
I was wondering whether this could be a warning or info rather than an error. Should we wait for a snapshot with a larger height in this case?

Reply:
We explicitly requested snapshot for this height though, so this would indicate a malicious/bugged lc contract.